// ng-carousel-cdk/esm2015/lib/carousel.component.js
import { ChangeDetectionStrategy, Component, ContentChild, Input, Output, ViewEncapsulation } from '@angular/core';
import { map } from 'rxjs/operators';
import { CarouselConfig } from './carousel-config';
import { CarouselSlideDirective } from './carousel-slide.directive';
import { IdGenerator } from './private/models/id-generator';
import { CarouselService } from './private/service/carousel.service';
import { ANIMATION_ID_GENERATOR, SLIDE_ID_GENERATOR } from './private/tokens';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './private/service/carousel.service';
import * as ɵngcc2 from './private/views/carousel-engine.component';
export function idGeneratorFactory() {
    return new IdGenerator();
}
/**
 * Defines carousel API to work with
 */
export class CarouselComponent {
    constructor(carousel) {
        this.carousel = carousel;
        this.itemIndexChange = this.carousel.carouselStateChanges()
            .pipe(map((state) => state.activeItemIndex));
    }
    set slideRef(newSlideRef) {
        this.carousel.setSlideTemplate(newSlideRef
            ? newSlideRef.templateRef
            : null);
    }
    set config(newConfig) {
        newConfig = new CarouselConfig(newConfig);
        this.carousel.setConfig(newConfig);
    }
    next() {
        this.carousel.next();
    }
    prev() {
        this.carousel.prev();
    }
    setIndex(newIndex) {
        this.carousel.setItemIndex(newIndex);
    }
    /**
     * Programmaticaly recalculates carousel position in case of
     * container size changes or other size interactions
     */
    recalculate() {
        this.carousel.recalculate();
    }
}
CarouselComponent.ɵfac = function CarouselComponent_Factory(t) { return new (t || CarouselComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CarouselService)); };
CarouselComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CarouselComponent, selectors: [["ng-carousel"]], contentQueries: function CarouselComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, CarouselSlideDirective, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.slideRef = _t.first);
    } }, inputs: { config: "config" }, outputs: { itemIndexChange: "itemIndexChange" }, exportAs: ["ngCarousel"], features: [ɵngcc0.ɵɵProvidersFeature([
            CarouselService,
            {
                provide: SLIDE_ID_GENERATOR,
                useFactory: idGeneratorFactory
            },
            {
                provide: ANIMATION_ID_GENERATOR,
                useFactory: idGeneratorFactory
            },
        ])], decls: 1, vars: 0, consts: [[1, "carousel-engine"]], template: function CarouselComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "carousel-engine", 0);
    } }, directives: [ɵngcc2.CarouselEngineComponent], styles: [".carousel-engine{display:block;overflow:hidden;width:100%}"], encapsulation: 2, changeDetection: 0 });
CarouselComponent.ctorParameters = () => [
    { type: CarouselService }
];
CarouselComponent.propDecorators = {
    slideRef: [{ type: ContentChild, args: [CarouselSlideDirective,] }],
    config: [{ type: Input }],
    itemIndexChange: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CarouselComponent, [{
        type: Component,
        args: [{
                selector: 'ng-carousel',
                template: "<carousel-engine class=\"carousel-engine\"></carousel-engine>\n",
                providers: [
                    CarouselService,
                    {
                        provide: SLIDE_ID_GENERATOR,
                        useFactory: idGeneratorFactory
                    },
                    {
                        provide: ANIMATION_ID_GENERATOR,
                        useFactory: idGeneratorFactory
                    },
                ],
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                exportAs: 'ngCarousel',
                styles: [".carousel-engine{display:block;overflow:hidden;width:100%}"]
            }]
    }], function () { return [{ type: ɵngcc1.CarouselService }]; }, { itemIndexChange: [{
            type: Output
        }], slideRef: [{
            type: ContentChild,
            args: [CarouselSlideDirective]
        }], config: [{
            type: Input
        }] }); })();
