// ng-carousel-cdk/esm2015/lib/carousel.module.js
import { A11yModule } from '@angular/cdk/a11y';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CarouselSlideDirective } from './carousel-slide.directive';
import { CarouselComponent } from './carousel.component';
import { PreventGhostClickDirective } from './prevent-ghost-click.directive';
import { FocusBlockDirective } from './private/directives/untabbable.directive';
import { CarouselEngineComponent } from './private/views/carousel-engine.component';
import * as ɵngcc0 from '@angular/core';
export class CarouselModule {
}
CarouselModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CarouselModule });
CarouselModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CarouselModule_Factory(t) { return new (t || CarouselModule)(); }, imports: [[
            CommonModule,
            A11yModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CarouselModule, { declarations: function () { return [CarouselComponent,
        CarouselSlideDirective,
        CarouselEngineComponent,
        FocusBlockDirective,
        PreventGhostClickDirective]; }, imports: function () { return [CommonModule,
        A11yModule]; }, exports: function () { return [CarouselComponent,
        CarouselSlideDirective,
        PreventGhostClickDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CarouselModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    A11yModule,
                ],
                declarations: [
                    CarouselComponent,
                    CarouselSlideDirective,
                    CarouselEngineComponent,
                    FocusBlockDirective,
                    PreventGhostClickDirective,
                ],
                exports: [
                    CarouselComponent,
                    CarouselSlideDirective,
                    PreventGhostClickDirective,
                ]
            }]
    }], null, null); })();
