// ng-carousel-cdk/esm2015/lib/prevent-ghost-click.directive.js
import { Directive, ElementRef, HostListener } from '@angular/core';
import { asyncScheduler } from 'rxjs';
import { HammerProviderService } from './private/service/hammer-provider.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './private/service/hammer-provider.service';
export class PreventGhostClickDirective {
    constructor(elementRef, hammer) {
        this.elementRef = elementRef;
        this.hammer = hammer;
        this.shouldPreventClick = false;
    }
    processClickEvent(event) {
        if (this.shouldPreventClick) {
            event.preventDefault();
            event.stopImmediatePropagation();
        }
    }
    ngOnInit() {
        this.listenPanEndEvents();
    }
    ngOnDestroy() {
        if (this.hammerManager) {
            this.hammerManager.destroy();
        }
    }
    listenPanEndEvents() {
        this.hammerManager = this.hammer.managerFor(this.elementRef.nativeElement);
        if (!this.hammerManager) {
            return;
        }
        this.hammerManager.on('panend pancancel', () => {
            this.shouldPreventClick = true;
            asyncScheduler.schedule(() => {
                this.shouldPreventClick = false;
            });
        });
    }
}
PreventGhostClickDirective.ɵfac = function PreventGhostClickDirective_Factory(t) { return new (t || PreventGhostClickDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.HammerProviderService)); };
PreventGhostClickDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: PreventGhostClickDirective, selectors: [["", "ngCarouselPreventGhostClick", ""]], hostBindings: function PreventGhostClickDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function PreventGhostClickDirective_click_HostBindingHandler($event) { return ctx.processClickEvent($event); });
    } } });
PreventGhostClickDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: HammerProviderService }
];
PreventGhostClickDirective.propDecorators = {
    processClickEvent: [{ type: HostListener, args: ['click', ['$event'],] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PreventGhostClickDirective, [{
        type: Directive,
        args: [{
                selector: '[ngCarouselPreventGhostClick]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.HammerProviderService }]; }, { processClickEvent: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
