// ng2-carouselamos/esm2015/lib/ng2-carouselamos.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng2-carouselamos.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Ng2CarouselamosComponent } from "./ng2-carouselamos.component";
import * as ɵngcc0 from '@angular/core';
export class Ng2CarouselamosModule {
}
Ng2CarouselamosModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2CarouselamosModule });
Ng2CarouselamosModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng2CarouselamosModule_Factory(t) { return new (t || Ng2CarouselamosModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2CarouselamosModule, { declarations: function () { return [Ng2CarouselamosComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [Ng2CarouselamosComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng2CarouselamosModule, [{
        type: NgModule,
        args: [{
                declarations: [Ng2CarouselamosComponent],
                imports: [CommonModule],
                exports: [Ng2CarouselamosComponent]
            }]
    }], null, null); })();
